javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,227 @@
+
+- The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have collisions before we grow them. Usually a good balance of `0.75 to 1` is used.
+
+### Computation complexity
+
+A hash map is very efficient in its insertion, retrieving and removing operations. That is because we use array indexes to do those operations. A hash map has an average case complexity of `O(1)` for the following methods:
+- Insertion
+- Retrieving
+- Deletion
+
+Assuming we have a good hash map written. The worst case of those operations would be `O(n)` and that happens when we have all our data hashes to the same exact bucket.
+
+The Growth of our hash map has the complexity of `O(n)` at all times.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Self restriction and limitation
+
+  Before we get started, we need to lay down some ground rules. As we all know JavaScript is a dynamic language the restriction we have in other languages does not exist in JavaScript for example if you create an array of size `16` to be your buckets size, nothing stopping you from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
+
+  1. When you create a function that takes a hash code value and returns the bucket index, use this snippet before the function returns to make sure it's accessing inbound array indexes only.
+
+      ```javascript
+      if (index < 0 || index >= buckets.length) {
+        throw new Error("Trying to access index out of bound");
+      }
+      ```
+
+  1. Inside the class DO NOT use Objects to store your key value pairs, And DO NOT use the array method `push` to insert your nodes into the buckets, always use indexes.
+
+#### Time to create a real hash map
+
+  1. Start by creating a HashMap class or factory function, up to you. Then proceed to create the following methods:
+
+  1. `hash` takes a value and produces a hash code with it. There are a lot of thought that goes into creating the most optimal hashing function, it is up to you if you want to do a research to find a good hashing function but do not dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes that will lead to too many collisions.
+
+  1. `set` takes two arguments, the first is a key and the second is a value that is assigned to this key.
+
+     - Remember to grow your buckets size when it needs to, by calculating the `load factor` and `capacity`.
```

Comment:
I think another good reminder here is to think about what needs to happen when the given key already exists in the hash map.

This can be a bit of a balance. I'm sure some things we want them to "discover" while doing the project, but with this, I think it's entirely plausible that a learner does the project without testing or thinking about what happens when you `set()` with a key that's already there. So I think it'd be good to call attention to this case.

Reply:
I put it as part of the lesson that if a key already present, then its value is overwritten. I can add another small reminder here, like the one I put for growing the table.